### ISSUE TITLE
how to  select cuda:4,5,6,7

### ISSUE THREAD

**Issue by yangqiheng2019** [axolotl-ai-cloud/axolotl#3615]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

I have 8 cuda ,but 0-3 are used . now I want to set cuda:4,5,6,7
wo set :export CUDA_VISIBLE_DEVICES=4,5,6,7 ,but its useless

base_model: /home/yangnan/models/Qwen3.5-9B
processor_type: AutoProcessor

# These 3 lines are required for vision/multimodal training
skip_prepare_dataset: true
remove_unused_columns: false
sample_packing: false

chat_template: qwen3_5
datasets:
  - path: /home/yangnan/yiwu_datasets_add_0422.jsonl
    type: chat_template
    split: train[:1%]

dataset_prepared_path: last_run_prepared
val_set_size: 0.0
output_dir: /home/yangnan/largeModel/code/axolotl-main/outputs/qwen3.5_9b

adapter: lora
lora_model_dir: /home/yangnan/largeModel/code/axolotl-main/outputs/qwen3.5_9b_lora

sequence_len: 8192
pad_to_sequence_len: false

lora_r: 32
lora_alpha: 16
lora_dropout: 0.05
# Targets the language model attention and MLP layers.
lora_target_modules:
  - q_proj
  - k_proj
  - v_proj
  - o_proj
  - down_proj
  - up_proj
  # Uncomment to also target the linear attention (GatedDeltaNet) projections:
  # - linear_attn.in_proj_qkv
  # - linear_attn.in_proj_z
  # - linear_attn.out_proj

wandb_project:
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 1
num_epochs: 5
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.0002

bf16: true
tf32: true

gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false
logging_steps: 1
flash_attention: true

warmup_ratio: 0.1
evals_per_epoch: 1
saves_per_epoch: 1
weight_decay: 0.0

### Current behaviour

no

### Steps to reproduce

no 

### Config yaml

```yaml

```

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by thad0ctor** [comment]
This works for me — I set CUDA_DEVICE_ORDER=PCI_BUS_ID alongside CUDA_VISIBLE_DEVICES and launch with torchrun directly (with --nproc_per_node   matching the number of visible GPUs). Excuse the extra logging/filtering bits  in my wrapper:                                                                
                                                                  
```
  export CUDA_DEVICE_ORDER=PCI_BUS_ID
  export CUDA_VISIBLE_DEVICES=0,3                                               
   
  script -qefc \                                                                
      "torchrun --nproc_per_node=2 --master_port=29500 -m axolotl.cli.train 
  \"$CONFIG_FILE\"" \                                                           
      "$RAW_TORCHRUN_LOG" 2>&1 \
      | python3 -u "$SCRIPT_DIR/filter_kwargs_warning.py"
```

Make sure everything is in the same shell and proc per node matches visible devices

**Comment by yangqiheng2019** [comment]
> This works for me — I set CUDA_DEVICE_ORDER=PCI_BUS_ID alongside CUDA_VISIBLE_DEVICES and launch with torchrun directly (with --nproc_per_node matching the number of visible GPUs). Excuse the extra logging/filtering bits in my wrapper:
> 
> ```
>   export CUDA_DEVICE_ORDER=PCI_BUS_ID
>   export CUDA_VISIBLE_DEVICES=0,3                                               
>    
>   script -qefc \                                                                
>       "torchrun --nproc_per_node=2 --master_port=29500 -m axolotl.cli.train 
>   \"$CONFIG_FILE\"" \                                                           
>       "$RAW_TORCHRUN_LOG" 2>&1 \
>       | python3 -u "$SCRIPT_DIR/filter_kwargs_warning.py"
> ```
> 
> Make sure everything is in the same shell and proc per node matches visible devices

THANK YOU VERY MUCH.
I  set export CUDA_VISIBLE_DEVICES=0,3   is useless , but i find " CUDA_VISIBLE_DEVICES=0,3   axolotl train  example/*****.yaml"   is useful.

**Comment by NanoCode012** [comment]
> I set export CUDA_VISIBLE_DEVICES=0,3 is useless , but i find " CUDA_VISIBLE_DEVICES=0,3 axolotl train example/*****.yaml" is useful.

Hello @yangqiheng2019 , thanks for the question. Admittedly I never ran export with cuda devices (only inline), though it's quite weird the former didn't work for you. Maybe it's some quotes? 

```
export CUDA_VISIBLE_DEVICES="0,3"  # does this work?
axolotl train example.yml
```

**Comment by reallyticsai** [comment]
The issue you're facing is likely due to the way CUDA devices are being managed in your environment. The `CUDA_VISIBLE_DEVICES` environment variable is used to specify which GPUs are visible to the process, but it doesn't control which GPUs are used by the training process itself. Here's a more detailed breakdown:

1. **CUDA_VISIBLE_DEVICES**: This variable is used to make specific GPUs available to the process. However, it doesn't necessarily mean that the training process will use those GPUs. It's more about making the GPUs visible to the process.

2. **GPU Selection**: When you run a training job, it typically uses the first available GPU in the list of visible GPUs, but it doesn't always respect the order specified in `CUDA_VISIBLE_DEVICES`. This could be why you're not seeing the desired GPUs being used.

3. **CUDA Runtime API**: To ensure that specific GPUs are used, you can use the CUDA Runtime API to explicitly select the GPUs. Here's an example of how you can modify your script to select GPUs 4, 5, 6, and 7:

```python
import os
import torch

# Set the CUDA_VISIBLE_DEVICES environment variable
os.environ["CUDA_VISIBLE_DEVICES"] = "4,5,6,7"

# Check if CUDA is available
if torch.cuda.is_available():
    print(f"Using GPU: {torch.cuda.get_device_name(0)}")
else:
    print("CUDA is not available")

# Now, you can use the GPUs explicitly
device = torch.device(f"cuda:{torch.cuda.current_device()}")
model.to(device)
```

4. **Training Script**: Ensure that your training script is set up to use the GPUs. You can use the `torch.cuda.device_count()` function to check how many GPUs are available

**Comment by yangqiheng2019** [comment]
> The issue you're facing is likely due to the way CUDA devices are being managed in your environment. The `CUDA_VISIBLE_DEVICES` environment variable is used to specify which GPUs are visible to the process, but it doesn't control which GPUs are used by the training process itself. Here's a more detailed breakdown:
> 
> 1. **CUDA_VISIBLE_DEVICES**: This variable is used to make specific GPUs available to the process. However, it doesn't necessarily mean that the training process will use those GPUs. It's more about making the GPUs visible to the process.
> 2. **GPU Selection**: When you run a training job, it typically uses the first available GPU in the list of visible GPUs, but it doesn't always respect the order specified in `CUDA_VISIBLE_DEVICES`. This could be why you're not seeing the desired GPUs being used.
> 3. **CUDA Runtime API**: To ensure that specific GPUs are used, you can use the CUDA Runtime API to explicitly select the GPUs. Here's an example of how you can modify your script to select GPUs 4, 5, 6, and 7:
> 
> import os
> import torch
> 
> # Set the CUDA_VISIBLE_DEVICES environment variable
> os.environ["CUDA_VISIBLE_DEVICES"] = "4,5,6,7"
> 
> # Check if CUDA is available
> if torch.cuda.is_available():
>     print(f"Using GPU: {torch.cuda.get_device_name(0)}")
> else:
>     print("CUDA is not available")
> 
> # Now, you can use the GPUs explicitly
> device = torch.device(f"cuda:{torch.cuda.current_device()}")
> model.to(device)
> 4. **Training Script**: Ensure that your training script is set up to use the GPUs. You can use the `torch.cuda.device_count()` function to check how many GPUs are available

wow，thanks very much，you are so nice.

**Comment by NanoCode012** [comment]
Does the above solve it for you?

**Comment by yangqiheng2019** [comment]
> Does the above solve it for you?

yes，the above  can solve it, you can try it.

**Comment by NanoCode012** [comment]
Thanks for confirming!
